consensus/src/lib.rs

```diff
@@ -64,7 +204,73 @@ where
 {
     match chain_config.net_upgrade().consensus_status(block_height) {
         RequiredConsensus::IgnoreConsensus => Ok(ConsensusData::None),
-        RequiredConsensus::PoS(_) => unimplemented!(),
+        RequiredConsensus::PoS(pos_status) => {
+            let pos_input_data = match input_data {
+                Some(GenerateBlockInputData::PoS(pos_input_data)) => pos_input_data,
+                Some(GenerateBlockInputData::PoW) => Err(ConsensusPoSError::PoWInputDataProvided)?,
+                None => Err(ConsensusPoSError::NoInputDataProvided)?,
+            };
+
+            let reward_destination = Destination::PublicKey(pos_input_data.stake_public_key());
+
+            let kernel_output = vec![TxOutput::ProduceBlockFromStake(
+                reward_destination.clone(),
+                pos_input_data.pool_id(),
+            )];
+
+            let block_reward = BlockRewardTransactable::new(
+                Some(pos_input_data.kernel_inputs()),
+                Some(&kernel_output),
+                None,
+            );
+
+            let kernel_input_utxos = pos_input_data.kernel_input_utxos();
+
+            let signature = StandardInputSignature::produce_uniparty_signature_for_input(
+                pos_input_data.stake_private_key(),
+                SigHashType::default(),
+                reward_destination,
+                &block_reward,
+                &kernel_input_utxos.iter().collect::<Vec<_>>(),
+                0,
+            )
+            .map_err(|_| ConsensusPoSError::FailedToSignKernel)?;
+
+            let input_witness = InputWitness::Standard(StandardInputSignature::new(
+                SigHashType::default(),
+                signature.encode(),
+            ));
+
+            let vrf_data = {
+                let sealed_epoch_randomness = sealed_epoch_randomness
+                    .ok_or(ConsensusPoSError::NoEpochData)?
+                    .randomness()
+                    .value();
+
+                let transcript = construct_transcript(
+                    chain_config.epoch_index_from_height(&block_height),
+                    &sealed_epoch_randomness,
+                    block_timestamp,
+                );
+
+                pos_input_data.vrf_private_key().produce_vrf_data(transcript.into())
+            };
+
+            let target_required = calculate_target_required_from_block_index(
+                chain_config,
+                &pos_status,
+                prev_block_index,
+                get_ancestor,
+            )?;
+
+            Ok(ConsensusData::PoS(Box::new(PoSData::new(
+                pos_input_data.kernel_inputs().clone(),
```

Comment:
can you calculate pool_id beforehand and avoid cloning kernel?

Reply:
Here, `pool_id` is just an accessor so there's no calculations being done. As for `kernel_inputs`, that accessor returns a ref which needs to be passed on - so it will either need to be `clone`d or we can keep in inside an `Option` and then `take()` it as a last step, but then that adds more steps each time it's accessed to see if it's empty.